### DISCUSSION TITLE
What is the right way to rate limit resources in pocketbase?

### DISCUSSION THREAD

**Discussion by xrd** [pocketbase/pocketbase#650] [Q&A] [answered]
I'm thinking about how I would implement limits to database tables (by logged in user). For example, imagine I want to permit each logged in user to create only 100 rows in a certain table. Is the best way to do this using a framework app? And, if so, what is the right method to implement (`OnModelBeforeCreate`)? Is there a good pattern to reject a creation event?

Or, is this something I can do with an API rule somehow? I tried to see if expanding the `@collection` fields would show me a count for a given record, but neither `length` nor `count` auto-populates, so I'm assuming this is not the way to go. And, I might want to have somewhat arbitrary and complex logic in here (such as, if the user is at "gold" level they can have 1000 records, whereas "silver" users only get 200 records, etc.)

**Comment by ganigeorgiev** [comment, marked as answer]
This cannot be done via an API rule, at least at the moment (with the support of computed fields in the near future per #311, you'll be able to have a read-only `COUNT` field and use it inside an API rule).

I think the best way will be to extend PocketBase and add a listener to the `OnRecordBeforeCreateRequest` hook. This hook is similar to the `OnModelBeforeCreate`, but:
- is triggered only when the create is performed via the API (aka. manual db changes doesn't trigger it)
- it has access to the request context, allowing you to get the current authenticated user by doing something like:
   ```go
	app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
		loggedUser, _ := e.HttpContext.Get(apis.ContextUserKey).(*models.User)
		if loggedUser == nil {
			return rest.NewUnauthorizedError("Requires user authentication.", nil)
		}

		log.Println(e.Record) // still unsaved

		// do something with the record or perform some additional db operations using app.Dao().
		//
		// return an error to abort the create operation

		return nil
	})
   ```
   You could find some query helpers in https://github.com/pocketbase/pocketbase/blob/master/daos/record.go#L75-L97.
   If you need a specific generic helper, please let me know and I'll consider adding it for the next v0.8.0 release.

_Another alternative approach, if you are not comfortable using Go, could be to proxy the request through a middleware server-side service (eg. a server-side action of a SSR app) and use the PocketBase web API as admin to perform your checks (in future I'll consider adding support for api keys since it was requested several times already, but that will be after the refactoring)._

**Reply by xrd** [reply]
I was thinking more about this. What do you think about using `TRIGGER` functions? 

I started investigating the listener approach. I love the extensibility of pocketbase and was successful in quickly getting an API webhook created despite not having written go in a long time. And, when I started down the road of implementing a limiter based on your suggestions, I got stuck because I didn't understand how to navigate the golang APi. Specifically, I was confused by how I should get a hold of the model object (the DB name), how to translate SQL (which I know) into golang (which I generally don't),  what the right way to reject the query would be, and how that would be received in the JS client, and how to do this in a performant way (this logic will be happening against every single request). 

So, I started looking into whether I could do this in SQLite itself, and triggers seem like the right way. And, I don't see any drawbacks to this approach: it looks performant, won't disturb pocketbase, albeit with a little extra management in that I will need to manually create it (though perhaps this could be added as a migration, not so familiar with how that works yet).

For example, i came up with this:

```
CREATE TRIGGER validate_words_limits
before insert on words
begin
select case when  ( select count(*) from words where owner = new.owner) > 50 then raise(ABORT, "too many") end; end;
```

This seems to be perfect for my needs, and it works:

```
sqlite> insert into words (contents, owner) values( 'ahoy, 'chris@foobar.com');
Error: too many
```

If I look at the client, it indeed blocks the request, although I do wish I could get that error somehow in my response so I could implement the right thing for the user. But, this is perfect for now.

![failed-to-create-record](https://user-images.githubusercontent.com/17064/192572409-c7ba0b54-fca4-4332-9b22-1e564aafa30e.png)

I have to say, I'm so thrilled by pocketbase and how fast these experiments go. I love that I can scp my PRODUCTION sqlite database down to my local environment, play with it in my local dev environment with real data, then run `.schema` on sqlite to see the FULL command to run inside the sqlite terminal on production. It's such an extraordinary development experience.

**Reply by ganigeorgiev** [reply]
Yep, a trigger is also an option, but note that it will be checked even when you try to insert a record as admin.

I personally would go with the hook because it is more flexible (_the performance difference would be insignificant; at the end it is just a single count and the go code doesn't do too much additional work_). The hook version of your trigger would look something like this (_I haven't tested it_):

```go
func findTotalWordsByOwner(dao *daos.Dao, collection *models.Collection, ownerId string) (int, error) {
	var total int

	err := dao.RecordQuery(collection).
		Select("count(*)").
		AndWhere(dbx.HashExp{"owner": ownerId}).
		Row(&total)

	return total, err
}

app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
	if e.Record.Collection().Name != "words" {
		return nil // no "words" record, continue normal execution...
	}

        // optionally, you could check if the request was from an admin and skip the count...
	if admin, _ := e.HttpContext.Get(apis.ContextAdminKey).(*models.Admin); admin != nil {
		return nil // words limit doesn't apply to records created from admins...
	}

	total, err := findTotalWordsByOwner(app.Dao(), e.Record.Collection(), e.Record.GetStringDataValue("owner"))
	if err != nil || total > 50 {
		return rest.NewBadRequestError("Cannot create more words.", err)
	}

	return nil
})
```

_The documentation for extending PocketBase is indeed currently lacking. More in-depth guides and examples how to "glue" the sub packages will be added in the near future with the new doc site redesign._

**Reply by xrd** [reply]
Regarding the check against the admin user, my trigger would not be reached until after the API Rules are hit, right? So, I could safely add a check in my trigger that looks to see if there is an owner defined, and if not (meaning it is an admin) then permit the insert to go through.

Thanks for the discussion and the code samples. That's very helpful.

**Reply by ganigeorgiev** [reply]
> Regarding the check against the admin user, my trigger would not be reached until after the API Rules are hit, right?

That is true for `update`. The `create` action has slightly different behavior because the rule requires a record to exist in the database. So on `create` we do the following:
1. "Dry Submit" the record (without files upload) in a transaction
2. Run the create API rule against the "dry submitted" record and rollback the transaction.
3. If the create rule fail - directly return an error (no api hooks are executed).
4. If the create rule succeed, trigger the OnRecordBeforeCreateRequest hook and perform a regular record save (including files upload).

I haven't tested it, but I think that the sql trigger will be called even when executed as part of uncommitted transactions (the trigger action will be rollbacked with the transaction rollback). Or in other words, I think your trigger will be called twice - once within the "dry submit" transaction right before checking the API rules (step 1) and again after the rules check and the "real" save (step 4).
